### PR TITLE
NTT columns batch

### PR DESCRIPTION
**Pull request by HadarIngonyama**
This PR adds the columns batch feature - enabling batch NTT computation to be performed directly on the columns of a matrix without having to transpose it beforehand, as requested in issue #264. 

Also some small fixes to the reordering kernels were added and some unnecessary parameters were removes from functions interfaces.

### CODE REVIEW

File: icicle/appUtils/ntt/tests/verification.cu [hunk 1]
```diff
@@ -19,29 +19,65 @@ typedef curve_config::scalar_t test_data;
 void random_samples(test_data* res, uint32_t count)
 {
   for (int i = 0; i < count; i++)
+    // res[i] = i==0 ? test_scalar::one() : test_scalar::zero();
```
Comment (yshekel) [inline icicle/appUtils/ntt/tests/verification.cu:22]: remove commented code. Also in other places

File: icicle/appUtils/ntt/tests/verification.cu [hunk 2]
```diff
@@ -19,29 +19,65 @@ typedef curve_config::scalar_t test_data;
...
   }
 }
 
+__global__ void copy_batch(test_scalar* in, test_scalar* out, int row_size, int column_size){
+  int tid = blockDim.x * blockIdx.x + threadIdx.x;
+  if (tid >= row_size * column_size) return;
+  out[tid] = in[tid];
+}
+
+__global__ void copy_batch2(test_scalar* in, test_scalar* out, int row_size, int column_size){
```
Comment (yshekel) [inline icicle/appUtils/ntt/tests/verification.cu:40]: why need 3 such functions? same for transpose. If you need them, please give clear names

File: icicle/appUtils/ntt/tests/vri2.cu [hunk 3]
```diff
@@ -0,0 +1,228 @@
+
```
Comment (yshekel) [inline icicle/appUtils/ntt/tests/vri2.cu]: why is this test needed in addition to the other one?

File: icicle/appUtils/ntt/thread_ntt.cu [hunk 4]
```diff
@@ -211,8 +212,20 @@ public:
     }
   }
 
+  __device__ __forceinline__ void loadGlobalDataBatched(
```
Comment (yshekel) [inline icicle/appUtils/ntt/thread_ntt.cu:215]: is that only for column batch? if yes, please rename to make to code clearer

File: wrappers/rust/icicle-core/src/ntt/mod.rs [hunk 5]
```diff
@@ -77,6 +77,8 @@ pub struct NTTConfig<'a, S> {
     pub coset_gen: S,
     /// The number of NTTs to compute. Default value: 1.
     pub batch_size: i32,
+    /// True if the NTTs are stored as the columns of a matrix and not as the rows.
```
Comment (yshekel) [inline wrappers/rust/icicle-core/src/ntt/mod.rs:80]: I would rephrase and say that when true, NTT is computed over columns.

File: icicle/appUtils/ntt/ntt.cu
Comment (yshekel) [inline icicle/appUtils/ntt/ntt.cu:1]: you need to make sure that if column batch is true then you choose the mixed-radix alg since radix2 does not support it, right?
Also note that it means that there are a few cases that are no supported (like 2^7) since mixed radix is not supported on those.

**Review by yshekel — CHANGES_REQUESTED** [review]
overall looks good. Wrote some comments, mainly to make the code cleaner.
Note that you must make sure that radix2 is never selected for column batch and it means that some cases are not supported (like 2^7) so I think you should log a clear error.

I think that we must have tests in CI before merging it, including small NTTs and the cases that are not supported to make sure it behaves as expected.

File: icicle/appUtils/ntt/tests/verification.cu [hunk 6]
```diff
@@ -141,22 +198,35 @@ int main(int argc, char** argv)
...
   CHK_IF_RETURN(benchmark(true /*=print*/, count));
 
+  if (COLUMNS_BATCH) {
+    transpose_batch<<<(NTT_SIZE * BATCH_SIZE + 256 - 1)/256,256>>>(GpuOutputNew, GpuScalarsTransposed, BATCH_SIZE, NTT_SIZE);
+    cudaDeviceSynchronize();
+    printf("trans err %d\n", cudaGetLastError());
+    CHK_IF_RETURN(
+      cudaMemcpy(GpuOutputNew, GpuScalarsTransposed, NTT_SIZE * BATCH_SIZE * sizeof(test_data), cudaMemcpyDeviceToDevice));
+    cudaDeviceSynchronize();
+    printf("copy err %d\n", cudaGetLastError());
```
Comment (jeremyfelder) [inline icicle/appUtils/ntt/tests/verification.cu:212]: Are these `cudaDeviceSynchronize` needed or for debug?

File: icicle/appUtils/ntt/tests/verification.cu [hunk 7]
```diff
@@ -141,22 +198,35 @@ int main(int argc, char** argv)
...
+  }
+
   // verify
   CHK_IF_RETURN(
     cudaMemcpy(CpuOutputNew.get(), GpuOutputNew, NTT_SIZE * BATCH_SIZE * sizeof(test_data), cudaMemcpyDeviceToHost));
   CHK_IF_RETURN(
     cudaMemcpy(CpuOutputOld.get(), GpuOutputOld, NTT_SIZE * BATCH_SIZE * sizeof(test_data), cudaMemcpyDeviceToHost));
+   cudaDeviceSynchronize();
+    printf("copy err %d\n", cudaGetLastError());
```
Comment (jeremyfelder) [inline icicle/appUtils/ntt/tests/verification.cu:221]: Remove

**Comment by ChickenLover** [comment]
@HadarIngonyama please change the branch to `main` as we don't use `main` anymore. Also, I think this will break golang bindings if we don't change the NTTConfig in golang accordingly

File: icicle/appUtils/ntt/tests/verification.cu (hunk 1 above)
Comment (HadarIngonyama) [reply icicle/appUtils/ntt/tests/verification.cu]: done

File: icicle/appUtils/ntt/tests/verification.cu (hunk 2 above)
Comment (HadarIngonyama) [reply icicle/appUtils/ntt/tests/verification.cu]: removed

File: icicle/appUtils/ntt/tests/vri2.cu (hunk 3 above)
Comment (HadarIngonyama) [reply icicle/appUtils/ntt/tests/vri2.cu]: removed

File: icicle/appUtils/ntt/thread_ntt.cu (hunk 4 above)
Comment (HadarIngonyama) [reply icicle/appUtils/ntt/thread_ntt.cu]: done

File: wrappers/rust/icicle-core/src/ntt/mod.rs (hunk 5 above)
Comment (HadarIngonyama) [reply wrappers/rust/icicle-core/src/ntt/mod.rs]: done

File: icicle/appUtils/ntt/ntt.cu
Comment (HadarIngonyama) [reply icicle/appUtils/ntt/ntt.cu:1]: done

File: icicle/appUtils/ntt/tests/verification.cu (hunk 6 above)
Comment (HadarIngonyama) [reply icicle/appUtils/ntt/tests/verification.cu]: removed

File: icicle/appUtils/ntt/tests/verification.cu (hunk 7 above)
Comment (HadarIngonyama) [reply icicle/appUtils/ntt/tests/verification.cu]: removed

**Comment by HadarIngonyama** [comment]
I've fixed all the above comments, please review

**Review by DmytroTym — APPROVED** [review]
Overall looks good. I can add columns batch mode into one of Rust tests as part of this PR (or else will do it later on). I also have two performance related questions:

1. When running `verification.cu` with different parameters, I see columns batch mode consistently outperforming normal batch mode. One example: NTT size is 2^20, batch size is 128, out of place, NN ordering. Columns batch takes ~70.2 ms. and normal batch ~94.7. The kernels are the same, most of the difference is in relative speed of the first two kernels: `batch_elementwise_mul_with_reorder_kernel` - 9.5 ms. vs. 14 ms. and especially `reorder_digits_inplace_and_normalize_kernel` - 10.4 ms. vs. 25.8 ms. Do you understand why this happens?
2. For smaller NTTs, radix2 is faster and while this might be offset by a larger batch size, when NTT size times batch size is ~2^16 or smaller, radix2 seems to be better. If such small sizes are relevant, I can add support for column batching for radix2. Though both @cliff0412 's use-case and danksharding are significantly larger.

**Comment by HadarIngonyama** [comment]
> Overall looks good. I can add columns batch mode into one of Rust tests as part of this PR (or else will do it later on). I also have two performance related questions:
> 
> 1. When running `verification.cu` with different parameters, I see columns batch mode consistently outperforming normal batch mode. One example: NTT size is 2^20, batch size is 128, out of place, NN ordering. Columns batch takes ~70.2 ms. and normal batch ~94.7. The kernels are the same, most of the difference is in relative speed of the first two kernels: `batch_elementwise_mul_with_reorder_kernel` - 9.5 ms. vs. 14 ms. and especially `reorder_digits_inplace_and_normalize_kernel` - 10.4 ms. vs. 25.8 ms. Do you understand why this happens?
> 2. For smaller NTTs, radix2 is faster and while this might be offset by a larger batch size, when NTT size times batch size is ~2^16 or smaller, radix2 seems to be better. If such small sizes are relevant, I can add support for column batching for radix2. Though both @cliff0412 's use-case and danksharding are significantly larger.

1. Not sure how exactly it improves the reordering, I think in general there is room for optimizing all the reordering kernels.
2. Yes, it wood be great if you could add support for radix2 NTT also.
I think @LeonHibnik  is working on adding a rust test, please coordinate with him.

**Comment by DmytroTym** [comment]
I wanted to add a new test initially as all our tests are really cluttered with logic. But `test_ntt_batched` is a really convenient place for this check. Also didn't have the time do the radix2 version so that's future work I guess.

**Review by LeonHibnik — APPROVED** [review]
